Detectors/ITSMFT/ITS/tracking/include/ITStracking/DBScan.h

```diff
@@ -0,0 +1,141 @@
+}
+
+template <typename T>
+void DBScan<T>::classifyVertices(const int nContributors)
+{
+  classifyVertices([nContributors](std::vector<o2::its::Edge>& edges) { return edges.size() == 0 ? 0 : edges.size() >= static_cast<size_t>(nContributors - 1) ? 2 : 1; },
+                   [](State& s1, State& s2) { return static_cast<int>(s1.second) > static_cast<int>(s2.second); });
+}
+
+template <typename T>
+void DBScan<T>::classifyVertices(std::function<unsigned char(std::vector<Edge>& edges)> classFunction)
+{
+  mClassFunction = classFunction;
+  const size_t size = { this->mVertices->size() };
+  mStates.resize(size);
+
+  if (!this->isMultiThreading()) {
+    for (size_t iVertex{ 0 }; iVertex < size; ++iVertex) {
+      mStates[iVertex] = std::make_pair<int, unsigned char>(iVertex, classFunction(this->getEdges()[iVertex]));
+    }
+  } else {
+    const size_t stride{ static_cast<size_t>(std::ceil(this->mVertices->size() / static_cast<size_t>(this->mExecutors.size()))) };
+    for (size_t iExecutor{ 0 }; iExecutor < this->mExecutors.size(); ++iExecutor) {
+      // We cannot pass a template function to std::thread(), using lambda instead
+      this->mExecutors[iExecutor] = std::thread(
+        [iExecutor, stride, this](const auto& classFunction) {
+          for (size_t iVertex{ iExecutor * stride }; iVertex < stride * (iExecutor + 1) && iVertex < this->mVertices->size(); ++iVertex) {
+            mStates[iVertex] = std::make_pair<int, unsigned char>(iVertex, classFunction(this->getEdges()[iVertex]));
+          }
+        },
+        mClassFunction);
+    }
+  }
+  for (auto&& thread : this->mExecutors) {
+    thread.join();
+  }
+}
+
+template <typename T>
+void DBScan<T>::classifyVertices(std::function<unsigned char(std::vector<Edge>&)> classFunction, std::function<bool(State&, State&)> sortFunction)
+{
+  classifyVertices(classFunction);
+  std::sort(mStates.begin(), mStates.end(), sortFunction);
+}
+
+template <typename T>
+std::vector<int> DBScan<T>::getCores()
+{
+  std::vector<State> cores;
+  std::vector<int> coreIndices;
+  std::copy_if(mStates.begin(), mStates.end(), std::back_inserter(cores), [](const State& state) { return state.second == 2; });
+  std::transform(cores.begin(), cores.end(), std::back_inserter(coreIndices), [](const State& state) -> int { return state.first; });
+  return coreIndices;
+}
+
+template <typename T>
+std::vector<std::vector<int>> DBScan<T>::computeClusters()
+{
+  std::vector<std::vector<int>> clusters;
+  std::vector<int> cores = getCores();
+  std::vector<unsigned char> usedVertices(this->mVertices->size(), false);
+
+  for (size_t core{ 0 }; core < cores.size(); ++core) {
+    if (!usedVertices[cores[core]]) {
+      std::vector<unsigned char> clusterFlags = this->getCluster(cores[core]);
+      std::transform(usedVertices.begin(), usedVertices.end(), clusterFlags.begin(), usedVertices.begin(), std::logical_or<>());
+      clusters.emplace_back(this->getClusterIndices(clusterFlags));
```

Comment:
To avoid intermediate vector creation in the ``Graph::getClusterIndices`` then copying it to ``clusters``, isn't it more efficient to do something like this?
````
clusters.emplace_back();
this->getClusterIndices(clusters.back(), clusterFlags);
````

Reply:
Indeed you have a point. Do you think something like
```clusters.emplace_back(std::move(this->getClusterIndices(clusterFlags)));```
might be viable as well?

Reply:
@mconcas : yes, this will work as well (avoiding actual moving of memory blocks)